File: client/java/src/main/java/com/be/mesh/client/prsim/Cluster.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.prsim;

import com.be.mesh.client.annotate.Index;
import com.be.mesh.client.annotate.MPI;
import com.be.mesh.client.annotate.SPI;

/**
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface Cluster {

    /**
     * Election will election leader of instances.
     */
    @MPI("mesh.cluster.election")
    byte[] election(@Index(0) byte[] buff);

    /**
     * IsLeader if same level.
     */
    @MPI("mesh.cluster.leader")
    boolean isLeader();

}

File: plugin/raft/cluster.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package raft

import (
	"context"
	"github.com/be-io/mesh/client/golang/cause"
	"github.com/be-io/mesh/client/golang/macro"
	"github.com/be-io/mesh/client/golang/prsim"
)

func init() {
	var _ prsim.Cluster = new(meshCluster)
	macro.Provide(prsim.ICluster, new(meshCluster))
}

const Name = "raft"

type meshCluster struct {
}

func (that *meshCluster) Att() *macro.Att {
	return &macro.Att{Name: Name}
}

func (that *meshCluster) Election(ctx context.Context, buff []byte) ([]byte, error) {
	return nil, cause.NotImplementError()
}

func (that *meshCluster) IsLeader(ctx context.Context) (bool, error) {
	return false, cause.NotImplementError()
}
